Add coordinate wrapping

ojeda-e opened the pull request:
In this PR I added PBC conditions. Changes in this PR fix #36 and possibly #22 

- [x] `self.ag.wrap()` added to`__init__` and ` _single_frame`.
- [x] Added tests for coordinate wrapping, including negative coordinates in x, y, and z, as suggested in #22 
- [x] Since PBC has an effect on the overall results, respective tests for `results.z_surface` and `results.mean` were updated.
- [x] I am not sure mentors would be interested in keeping previous tests, so I didn't delete but marked them as `xfail(reason="PBC conditions not applied.")`  (See line 276 is `tests_membrane_curvature.py`.

@orbeckst @lilyminium @IAlibay are there any other relevant tests I am missing here?
Thanks

pep8speaks commented:
Hello @ojeda-e! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`membrane_curvature/base.py`](https://github.com/MDAnalysis/membrane-curvature/blob/1a436d8480ade67daec0c7df0b72eb2fb50bbac9/membrane_curvature/base.py):

> [Line 151:21](https://github.com/MDAnalysis/membrane-curvature/blob/1a436d8480ade67daec0c7df0b72eb2fb50bbac9/membrane_curvature/base.py#L151): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to True should be 'if cond is True:' or 'if cond:'

tylerjereddy commented on membrane_curvature/base.py:
```diff
@@ -146,6 +147,14 @@ def __init__(self, universe, select='all',
                 logger.warn(msg)
 
+        # Apply PBC conditions
+        if self.pbc is True:
```
Any reason for the more verbose `is True` check here vs. conventional `if self.pbc:`? You do that as expected below, for example.

ojeda-e replied:
Thanks for your comment. No, not really, I just learned from PEP8 the way you suggest is the best one. :)

richardjgowers commented on membrane_curvature/base.py:
```diff
@@ -17,8 +17,9 @@
 
 import MDAnalysis
 from MDAnalysis.analysis.base import AnalysisBase
+from MDAnalysis.transformations import wrap
```
This isn't being used as far as I can see? The wrap method on AtomGroups requires no import.

lilyminium commented:
I might be missing something, but could you explain how coordinate wrapping treats PBC? As far as I can tell you're still not treating the boundary rows and columns of the surface grid?

I'm also not convinced that you can validly wrap atom coordinates when they have been rotated and translated as I believe you preprocess your trajectories to do. Perhaps you could instead get the translation, or translation + rotation matrix for each frame, and use that to transform your surface?

This is indirectly related to @orbeckst's project 5 of GSoC (https://github.com/MDAnalysis/mdanalysis/wiki/Project-Ideas-2021#project-5-general-unit-cell-representation) but just storing the matrix for each frame of this analysis would be a good starting point if you think doing the whole thing is too out of scope.

ojeda-e commented:
I pushed changes to use `wrap` instead of PBC. I didn't push changes for the keyword that will discriminate between raw and processed trajectories, as we discussed in this [comment](https://github.com/MDAnalysis/membrane-curvature/issues/36#issuecomment-882280526). 
To clarify, should I add such changes in this PR as well @lilyminium @orbeckst @IAlibay ?

ojeda-e commented:
@orbeckst @lilyminium @IAlibay @richardjgowers do I have a green light here? Thanks.

richardjgowers commented on membrane_curvature/base.py:
```diff
@@ -35,8 +35,8 @@ class MembraneCurvature(AnalysisBase):
-    pbc : bool, optional
-        Apply periodic boundary conditions.
+    wrap : bool, optional
+        Apply coordinate wrapping.
```
I'd clarify what this means.  This will pack all the particles into the primary unit cell before processing them further right?

ojeda-e replied:
Yes, correct, it's to keep all articles in the primary unit cell. I'll update this. Thanks.

richardjgowers commented:
@ojeda-e looks good but I'd clarify what wrapping is doing here.  PBC treatment is confusing as it means one of two very different things

orbeckst commented on membrane_curvature/base.py:
```diff
@@ -146,6 +146,14 @@ def __init__(self, universe, select='all',
+        if self.wrap:
+            self.ag.wrap()
+        else:
+            warnings.warn(" `wrap == False` may result in inaccurate calculation "
```
ALso send to logger.warning

orbeckst reviewed (commented):
I would also add some documentation explaining when you should use `wrap=True` and explain in words what it achieves.

ojeda-e commented:
> I would also add some documentation explaining when you should use `wrap=True` and explain in words what it achieves.

Thanks, @orbeckst, initially I was planning to add the explanation of when to use `wrap=True` in the [Usage](https://membrane-curvature.readthedocs.io/en/latest/source/pages/Usage.html) (Issue #58)  

Now that you point it out I'll add a brief description here in `base.py` and a more detailed one in [Usage](https://membrane-curvature.readthedocs.io/en/latest/source/pages/Usage.html).

ojeda-e commented:
Can I have an approval here, please? @richardjgowers @orbeckst @lilyminium

lilyminium commented on membrane_curvature/base.py:
```diff
@@ -128,6 +142,16 @@ def __init__(self, universe, select='all',
+        if self.wrap:
+            self.ag.wrap()
+        else:
+            msg = (" `wrap == False` may result in inaccurate calculation "
+                   "of membrane curvature. Surfaces will be derived from "
+                   "a reduced number of atoms.")
+            warnings.warn(msg)
+            logger.warn(msg)
```
Can you please add in something about ignoring this warning if the trajectory is processed to rotate -- otherwise, people who are following the directions and not using `wrap=True` will be unnecessarily alarmed.

lilyminium commented on membrane_curvature/base.py:
```diff
@@ -68,8 +68,20 @@ class MembraneCurvature(AnalysisBase):
+
     Notes
     -----
+    Use `wrap=True` when `mda.Universe` contains the raw trajectory, and
+    `wrap=False` for processed trajectories. For more details on when to use
+    `wrap=True`, check the `Usage
+    <https://membrane-curvature.readthedocs.io/en/latest/source/pages/Usage.html>`_
+    page.
```
Can this be a little more fine-grained -- do many engines unwrap atoms in raw trajectories? If the trajectory is raw, it's usually more likely that the atoms are already in the unit cell, right? Could you mention translation vs rotation?

ojeda-e replied:
I changed the lines a bit. Happy to received suggestions for the pushed changes.

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -2,6 +2,7 @@
 Unit and regression test for the membrane_curvature package.
 """
 
+from numpy.lib.stride_tricks import DummyArray
```
What's this for?

ojeda-e replied:
Sorry, I have no idea how that line ended up there.

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -286,24 +299,82 @@ def test_analysis_get_z_surface(self, universe, x_bin, y_bin, expected_surface):
         avg_surface = mc.results.average_z_surface
         assert_almost_equal(avg_surface, expected_surface)
 
-    def test_analysis_mean(self, universe):
```
Instead of deleting this test and replacing it with the  new one, could you please keep both tests for `wrap=True` and `wrap=False`?

ojeda-e replied:
Ok, I added a test for both. Thanks for this. :)

lilyminium reviewed (changes requested):
@ojeda-e just a few suggestions :-)

ojeda-e commented on membrane_curvature/data/test_po4_small.gro:
```diff
@@ -9,4 +9,4 @@ Test file 9 lipids in grid
  2713POPC   PO4    8   1.000   2.000  12.000
  2999POPC   PO4    9   2.000   2.000  12.000
-  3.00000  3.00000  3.00000   
+  3.00000  3.00000  30.00000   
```
I'm sorry for not noticing this before. The fact that the box has `u.dimension[2]=30 A` (or 3nm), made the tests in lines 303 and 315 fail. Since 120 is a multiple of 30, the `derived_surface` was zero and therefore, everything else was zero.  Hope it makes sense.

lilyminium replied:
Ah, thanks for that explanation, I wouldn't have known where to look otherwise!

ojeda-e commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -211,7 +212,7 @@ def test_get_z_surface(x_bin, y_bin, x_range, y_range, expected_surface):
 class TestMembraneCurvature(object):
     @pytest.fixture()
     def universe(self):
-        return mda.Universe(GRO_PO4_SMALL, XTC_PO4_SMALL)
```
I deleted the trajectory for now since I would need to create a new trajectory test with `u.dimensions[2]=30`. Skipping the trajectory the tests pass.

ojeda-e commented:
I discover a "little" bug in my `test_po4_small.gro` (see [comment](https://github.com/MDAnalysis/membrane-curvature/pull/48/files#r680270555)). Given that the size of the box in `z` was 30 A, `ag.wrap` was returning `z=0` (and therefore everything was zero, surface, mean, Gaussian). I also deleted the trajectory test for now, as I mentioned in this [comment](https://github.com/MDAnalysis/membrane-curvature/pull/48/files#r680271329). I don't think is strictly necessary, but if you think is best to add the traj, I will add it. I hope we all agree this change doesn't represent any harm or has a negative impact in the tests. 

Thanks to @lilyminium for asking me to add tests with `wrap=False` and `wrap=True` because otherwise, I'd have never seen that bug.   :)

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -297,13 +311,96 @@ def test_analysis_mean(self, universe):
+                               select='name PO4',
+                               n_x_bins=3,
+                               n_y_bins=3,
+                               wrap='False').run()
```
```suggestion
                               wrap=False).run()
```

This string will still evaluate truthily, i.e. wrapping is still turned on. So, it's hard for this test to evaluate whether `wrap=False` is working. Switching to a new universe (maybe with dummy coordinates if that's easiest) where `wrap=True` and `wrap=False` give you different results might be the best way?

ojeda-e replied:
Sorry about that and thanks for spotting it.

ojeda-e commented on membrane_curvature/surface.py:
```diff
@@ -87,6 +87,10 @@ def get_z_surface(coordinates, n_x_bins=10, n_y_bins=10, x_range=(0, 100), y_ran
     for l, m, z in zip(cell_x_floor, cell_y_floor, z_coords):
 
         try:
+            if l < 0 or m < 0:
```
without this condition, negative indexes will pass regardless because Python returns the i-th element `(-1, 0) = (i, 0)`. So previously, all the negative indexes were in fact passed.

lilyminium replied:
If you warn you can also add a test that it happens!

ojeda-e commented:
I added two fixtures for universes:
- `universe_dummy_wrap` with atoms out of bounds in `x`,
- `universe_dummy_wrap_xy` with atoms out of bounds in `x` and `y`.

Intentionally, I made the two fixtures match with the derived `z_surface` when `wrap=True` is applied. Then, the third fixture added `dummy_surface`. For each one of the universes with atoms out of bounds, I added a test for
- .results.average_surface
- .results.average_mean
- .results.average_gaussian.

For each test that involves curvature, I calculated `mean_curvature` and `gaussian_curvature` directly from `dummy_surface`. 

I also discover another bug in `surface.py` (see [comment](https://github.com/MDAnalysis/membrane-curvature/pull/48/files#r680425288)). But fixed it with an if statement. Hope you are ok with this change as well.

Thanks again @lilyminium

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -220,10 +222,59 @@ def universe_dummy(self):
+        #   +-----------+
+        #   |   | 1 | 2 | 3
+        #   +-----------+
+        a = np.array([[300., 0., 110.], [100., 0., 150.], [200., 0., 150.],
+                      [0., 100., 150.], [100., 100., 150.], [-100., 100., 150.],
+                      [300., 200., 110.], [100., 200., 150.], [200., 200., 150.]])
```
I'm not sure this corresponds to the picture (assuming it goes 1 to 9). Shouldn't 1 wrap from `[300, 0, 110]` to `[0, 0, 110]`? If so, I don't understand how either the unwrapped or wrapped version of atom 1 is in the center of the bottom row?

```python
>>> import MDAnalysis as mda
>>> import numpy as np
>>> a = np.array([[300., 0., 110.], [100., 0., 150.], [200., 0., 150.],
...               [0., 100., 150.], [100., 100., 150.], [-100., 100., 150.],
...               [300., 200., 110.], [100., 200., 150.], [200., 200., 150.]])
>>> u = mda.Universe(a, n_atoms=9)
>>> u.dimensions = [300, 300, 300, 90, 90, 90]
>>> ag2 = u.atoms.wrap(inplace=False)
>>> ag2
array([[  0.,   0., 110.],
       [100.,   0., 150.],
       [200.,   0., 150.],
       [  0., 100., 150.],
       [100., 100., 150.],
       [200., 100., 150.],
       [  0., 200., 110.],
       [100., 200., 150.],
       [200., 200., 150.]], dtype=float32)
```

ojeda-e replied:
Yes, the indexes in the picture were not accurate. Updated.

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -220,10 +222,59 @@ def universe_dummy(self):
+
+        return u
+
+    # Equivalent to universe_dummy_wrap when wrapping is applied.
+    # Atoms out of bounds in x and y
+    @pytest.fixture()
+    def universe_dummy_wrap_xy(self):
+        a = np.array([[300., 0., 110.], [100., 0., 150.], [200., 0., 150.],
+                      [0., -200., 150.], [100., -200., 150.], [-100., 100., 150.],
+                      [300., 200., 110.], [100., 200., 150.], [200., 200., 150.]])
```
The 300s will convert to 0 unless you change the bounds to 300.01 or similar, so I don't think it'll be equivalent? Also, suggestion: would `dummy_universe_unwrapped_xy` be better, because this is *not* wrapped?

lilyminium replied:
I'm not quite sure what this does that `universe_dummy_wrap` does not, actually, since `universe_dummy_wrap` also wraps to different coordinates?

ojeda-e replied:
Sorry if I wasn't clear. Both universes are unwrapped. When they wrap, they are equivalent. Then there is a test where we can check that coordinates wrap correctly in x, and another separate that wraps correctly in x and y.
```
import MDAnalysis as mda
import numpy as np

array_x = np.array([[300., 0., 110.], [100., 0., 150.], [200., 0., 150.],
              [0., 100., 150.], [100., 100., 150.], [-100., 100., 150.],
              [300., 200., 110.], [100., 200., 150.], [200., 200., 150.]])

u = mda.Universe(array_x, n_atoms=9)
u.dimensions = [300, 300, 300,  90., 90., 90.]
ag2 = u.atoms.wrap()
ag2
array([[  0.,   0., 110.],
       [100.,   0., 150.],
       [200.,   0., 150.],
       [  0., 100., 150.],
       [100., 100., 150.],
       [200., 100., 150.],
       [  0., 200., 110.],
       [100., 200., 150.],
       [200., 200., 150.]], dtype=float32)
```
and
 
```
array_xy = np.array([[300., 0., 110.], [100., 0., 150.], [200., 0., 150.],
              [0., -200., 150.], [100., -200., 150.], [-100., 100., 150.],
              [300., 200., 110.], [100., 200., 150.], [200., 200., 150.]])

u = mda.Universe(array_xy, n_atoms=9)
u.dimensions = [300, 300, 300,  90., 90., 90.]
ag2_xy = u.atoms.wrap()
ag2_xy

ag2_x == ag2_xy
array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])
```
If that's ok, I'd prefer to keep both cases. I think it's clearer. In the first case, only atoms are out of bounds in the x coordinate, while in the second atoms are out of bounds in x and y.

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -297,13 +350,238 @@ def test_analysis_mean(self, universe):
+                   [150., 120., 120., np.nan],
+                   [np.nan, np.nan, np.nan, np.nan]])),
+        (5, 5,
+         np.array([[150., 120., np.nan, 150., np.nan],
+                  [150., 120., np.nan, 120., np.nan],
+                  [np.nan, np.nan, np.nan, np.nan, np.nan],
+                  [150., 120., np.nan, 120., np.nan],
+                  [np.nan, np.nan, np.nan, np.nan, np.nan]]))
```
Why are the nans now in the 3rd row and column? Should it not be like the 4x4 grid but with nans in the extra rows and columns? I don't think this is the answer we'd want to get for the surface.

lilyminium replied:
Ah, I forgot that xrange and yrange defaulted to the universe dimensions. Never mind.

ojeda-e replied:
This test runs with the grofile with 9 particles (`GRO_PO4_SMALL`). In the file, the 9 particles in the system are distributed like this:
```
   +-----------+
   | 7 | 8 | 9 | 
   +-----------+
   | 4 | 5 | 6 |
   +-----------+
   | 1 | 2 | 3 | 
   +-----------+
```
, where atoms 1-4 have `z=150`, and atoms 5-9 have `z=120`. 
Then, when we divide the same grid into 4 bins, we have that each bin covers a region of 0.75

```
  y = 3+---------------+
       | x | x | x | x | 
       +---------------+
       | 7 | 8 | 9 | x | 
       +---------------+
       | 4 | 5 | 6 | x | 
       +---------------+
       | 1 | 2 | 3 | x |
 y = 0 +---------------+
      +---|---|---|----+ 
       x=0             x=3
```
With 5 bins, we have that each bin covers a region of 0.6. 
Atom 3= [ 20.,   0., 150.], --> `1.8 < x < 2.4`, `0. < y < 0.6`,
Atom 6= [ 20.,   10., 150.], --> `1.8 < x < 2.4`, `0.6 < y < 1.2`
Atom 7 = [  0.,  20., 120.], --> `0. < x < 0.6`, `1.8 < y < 2.4`
Atom 8=  [ 10.,  20., 120.], --> `0.6 < x < 1.2`, `1.8 < y < 2.4`
Atom 9 = [ 20.,  20., 120.] --> `1.8 < x < 2.4`, `1.8 < y < 2.4`
then all the space between `1.8 < x < 2.4` and 1.8 < y < 2.4` is empty.
```
y=3 +-------------------+
    | x | x | x | x | x | 
2.4 +-------------------+
    | 7 | 8 | x | 9 | x | 
1.8 +-------------------+
    | x | x | x | x | x |
1.2 +-------------------+
    | 4 | 5 | x | 6 | x |
0.6 +-------------------+
    | 1 | 2 | x | 3 | x |
y=0 +-------------------+
    +---|---|---|---|---+ 
   x=0 0.6 1.2 1.8 2.4 x=3
```
This is why the 3rd row and column of nans. Sorry it took forever to answer.

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -297,13 +350,238 @@ def test_analysis_mean(self, universe):
\ No newline at end of file
+        assert_almost_equal(avg_mean, expected_mean)
+
+    @pytest.mark.parametrize('dummy_array, expected_surface', [
+        # test with negative x coordinates
+        (np.array([[0., 0., 150.], [-100., 0., 150.], [- 200., 0., 150.],
+                  [0., 100., 150.], [-100., 100., 120.], [-200., 100., 120.],
+                  [0., 200., 120.], [-100., 200., 120.], [-200., 200., 120.]]),
+         np.array([[150., 150., 120.],
+                   [150., 120., 120.],
+                   [150., 120., 120.]])),
+        # test with negative y coordinates
+        (np.array([[0., 0., 150.], [100., 0., 150.], [200., 0., 150.],
+                  [0., -100., 150.], [100., -100., 120.], [200., -100., 120.],
+                  [0., -200., 120.], [100., -200., 120.], [200., -200., 120.]]),
+            np.array([[150., 120., 150.],
+                      [150., 120., 120.],
+                      [150., 120., 120.]])),
+        # test with negative z coordinates
+        (np.array([[0., 0., -150.], [100., 0., -150.], [200., 0., -150.],
+                  [0., 100., -150.], [100., 100., 120.], [200., 100., 120.],
+                  [0., 200., 120.], [100., 200., 120.], [200., 200., 120.]]),
+            np.array([[150., 150., 120.],
+                      [150., 120., 120.],
+                      [150., 120., 120.]]))
```
If all of these unwrap to the same coordinates, I think just the tests you already have above for the surface are enough for the x-y and you could delete the ones here. The z-unwrapping is a good thought though!

ojeda-e replied:
Ok, thanks. I'll leave z coordinates.

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -297,13 +350,238 @@ def test_analysis_mean(self, universe):
+    # test mean curvature
+    def test_analysis_mean_wrap(self, universe_dummy_wrap, dummy_surface):
+        x_bin = y_bin = 3
+        mc = MembraneCurvature(universe_dummy_wrap,
+                               n_x_bins=x_bin,
+                               n_y_bins=y_bin).run()
```
Seeing as you repeat the same analysis over and over, a fixture might be a good choice to save time.

```python
@pytest.fixture()
def wrapped_curvature_of_unwrapped_universe(universe_dummy_wrap):
	return MembraneCurvature(universe_dummy_wrap,
                               n_x_bins=3,
                               n_y_bins=3).run()

def test_analysis_mean_wrap(self, wrapped_curvature_of_unwrapped_universe, dummy_surface):
	avg_mean = wrapped_curvature_of_unwrapped_universe.results.average_mean
    expected_mean = mean_curvature(dummy_surface)
    assert_almost_equal(avg_mean, expected_mean)
```

lilyminium commented:
Thank you for adding the new tests @ojeda-e!

ojeda-e commented:
The pushed changes include:

- [x] Corrected indexes in diagram:
```
    #   +-----------+          +-----------+
    #   |   | 8 | 9 | 7        | 7 | 8 | 9 |
    #   +-----------+          +-----------+
    # 6 | 4 | 5 |   |   --->   | 4 | 5 | 6 |
    #   +-----------+          +-----------+
    #   |   | 2 | 3 | 1        | 1 | 2 | 3 |
    #   +-----------+          +-----------+
    #
```

- [x] Added fixtures that returns membranecurvature.run() for each universe with unwrapped coordinates:
   - `curvature_unwrapped_universe(self, universe_dummy_wrap)`
   - `curvature_unwrapped_universe_xy(self, universe_dummy_wrap_xy)`

- [x] Removed previous tests for x and y coordinates as [suggested](https://github.com/MDAnalysis/membrane-curvature/pull/48#discussion_r680560734), and added an extra one with negative z coordinates with 4 bins.

- [x] In `surface.py`, prints were replaced by warnings as [suggested](https://github.com/MDAnalysis/membrane-curvature/pull/48#discussion_r680557817). However I have a question here:
   - I understand the purpose of logger in `base.py`.  But for the warning in `surface.py`, I am not sure I should add a logger. Would it make sense to add a logger there?

Thanks @lilyminium

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -366,20 +377,16 @@ def test_analysis_mean_no_wrap(self, universe):
+        avg_surface = curvature_unwrapped_universe.results.average_z_surface
         assert_almost_equal(avg_surface, dummy_surface)
 
     # test surface in universe with atoms out of bounds in x and y
```
You could use the `curvature_unwrapped_universe_xy` fixture here too, right? (the line after)

ojeda-e replied:
Oh, yeah, I missed that, thanks!

lilyminium reviewed (approved):
Thanks @ojeda-e, the new indices are much easier to understand! I found another place you could use a fixture, but otherwise, this looks good to me :D

> Would it make sense to add a logger there?

This might come down to personal preference. I don't ever log warnings because users can just ``logging.captureWarnings(True)`` in their scripts if they really want it to log. @orbeckst might have a different opinion.

ojeda-e commented:
Thanks for checking this @lilyminium! This was a very buggy PR 🐛  but glad it's all fixed now! :D
